### PR TITLE
feat: Add support for provider events.

### PR DESCRIPTION
**Pull request by kinyoklion**

### CODE REVIEW

File: ld_openfeature/provider.py [hunk 1]
```diff
@@ -29,51 +85,52 @@ def get_provider_hooks(self) -> List[Hook]:
         return []
 
     def resolve_boolean_details(
-        self,
-        flag_key: str,
-        default_value: bool,
-        evaluation_context: Optional[EvaluationContext] = None,
+            self,
```
Comment (kinyoklion) [inline ld_openfeature/provider.py:88]: Auto formatting did this. I could undo them.

File: ld_openfeature/provider.py [hunk 2]
```diff
@@ -19,7 +22,60 @@ def __init__(self, config: Config):
...
+            error_message = self.__get_message(status, "encountered an unknown error")
+            self.emit_provider_stale(ProviderEventDetails(message=error_message))
+
+        # For now treat an unknown state as no change.
+
+    def __handle_flag_change(self, change: FlagChange):
+        self.emit_provider_configuration_changed(ProviderEventDetails(flags_changed=[change.key]))
+        pass
+
+    def initialize(self, evaluation_context: EvaluationContext):
```
Comment (kinyoklion) [inline ld_openfeature/provider.py:46]: This should handle the cases where the client failed or succeeded within the start_wait period, or where it exceeded that period and the data source later becomes VALID or OFF.

File: tests/test_provider.py [hunk 3]
```diff
@@ -128,3 +135,118 @@ def test_logger_changes_should_cascade_to_evaluation_converter(provider: LaunchD
...
+
+    api.add_handler(ProviderEvent.PROVIDER_ERROR, handle_status)
+
+    openfeature_provider = LaunchDarklyProvider(
+        Config("", update_processor_class=FailingDataSource, send_events=False))
+
+    try:
+        api.set_provider(openfeature_provider)
+        # The method should throw
+        assert False
```
Comment (keelerm84) [inline tests/test_provider.py:183]: I think you can also do this, which is probably slightly preferred. 
```suggestion
        pytest.fail("set_provider should have raised an exception")
```

File: tests/test_provider.py [hunk 4]
```diff
@@ -128,3 +135,118 @@ def test_logger_changes_should_cascade_to_evaluation_converter(provider: LaunchD
...
+    except Exception:
+        pass
+
+    with lock:
+        assert ld_provider_error_count == 1
+    pass
```
Comment (keelerm84) [inline tests/test_provider.py:189]:
```suggestion
```

File: tests/test_provider.py [hunk 5]
```diff
@@ -128,3 +135,118 @@ def test_logger_changes_should_cascade_to_evaluation_converter(provider: LaunchD
...
+    api.add_handler(ProviderEvent.PROVIDER_ERROR, handle_status)
+
+    openfeature_provider = LaunchDarklyProvider(
+        Config("", update_processor_class=DelayedFailingDataSource, send_events=False))
+
+    api.set_provider(openfeature_provider)
+
+    with lock:
+        assert ld_provider_error_count == 1
+    pass
```
Comment (keelerm84) [inline tests/test_provider.py:214]:
```suggestion
```

File: ld_openfeature/provider.py (hunk 1 above)
Comment (keelerm84) [reply ld_openfeature/provider.py:88]: If you don't mind. That feels really over-indented to me.

**Review by keelerm84 — APPROVED** [review]

File: tests/test_provider.py (hunk 4 above)
Comment (kinyoklion) [reply tests/test_provider.py:189]: Oops. I tend to write in all the functions I need with `pass` and then sometimes maybe forget to remove them.

**Pull request merged**